sys/include/net/sock/dtls.h

```diff
@@ -0,0 +1,557 @@
+ * sock_dtls_destroy(&dtls_sock);
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * @{
+ *
+ * @file
+ * @brief   DTLS sock definitions
+ *
+ * @author  Aiman Ismail <muhammadaimanbin.ismail@haw-hamburg.de>
+ */
+
+#ifndef NET_SOCK_DTLS_H
+#define NET_SOCK_DTLS_H
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <sys/types.h>
+
+#include "net/sock.h"
+#include "net/sock/udp.h"
+#include "net/credman.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Type for a DTLS sock object
+ *
+ * @note    API implementors: `struct sock_dtls` needs to be defined by
+ *          implementation-specific `sock_dtls_types.h`.
+ */
+typedef struct sock_dtls sock_dtls_t;
+
+/**
+ * @brief Information about the established session with
+ *        the remote endpoint. Used when sending and
+ *        receiving data to the endpoint
+ */
+typedef struct sock_dtls_session sock_dtls_session_t;
+
+/**
+ * @brief Must be called once before any other use
+ */
+void sock_dtls_init(void);
+
+/**
+ * @brief Creates a new DTLS sock object
+ *
+ * @param[out] sock     The resulting DTLS sock object
+ * @param[in] udp_sock  Existing UDP sock to be used
+ * @param[in] tag       Credential tag of the sock. Used to get the right
+ *                      credential from pool.
+ * @param[in] method    Defines the method for the client or server to use.
+ *
+ * @return  0 on success.
+ * @return  value < 0 on error
+ */
+int sock_dtls_create(sock_dtls_t *sock, sock_udp_t *udp_sock,
+                     credman_tag_t tag,unsigned method);
+
+/**
+ * @brief Initialises the server to listen for incoming connections
+ *
+ * @param[in] sock      DTLS sock to listen to
+ */
+void sock_dtls_init_server(sock_dtls_t *sock);
+
+/**
+ * @brief Establish DTLS session to a server. Execute the handshake step in
+ *        DTLS.
+ *
+ * @param[in]  sock     DLTS sock to use
+ * @param[in]  ep       Endpoint to establish session with
+ * @param[out] remote   The established session, cannot be NULL
+ *
+ * @return  0 on success
+ * @return  value < 0 on other errors
+ * @return  -EAGAIN, if @p timeout is `0` and no data is available.
+ * @return  -EADDRNOTAVAIL, if the local endpoint of @p sock is not set.
+ * @return  -EINVAL, if @p remote is invalid or @p sock is not properly
+ *          initialized (or closed while sock_udp_recv() blocks).
+ * @return  -ENOBUFS, if buffer space is not large enough to store received
+ *          credentials.
+ * @return  -ETIMEDOUT, if timed out when trying to establish session.
+ */
+int sock_dtls_establish_session(sock_dtls_t *sock, sock_udp_ep_t *ep,
+                                sock_dtls_session_t *remote);
+
+/**
+ * @brief Close an existing DTLS session
+ *
+ * @pre `(sock != NULL) && (ep != NULL)`
+ *
+ * @param[in] sock      @ref sock_dtls_t, which the session is established on
+ * @param[in] remote    Remote session to close
+ */
+void sock_dtls_close_session(sock_dtls_t *sock, sock_dtls_session_t *remote);
```

Comment:
Not to be confused with `sock_dtls_close()` (in case we rename `sock_dtls_destroy()`) how about naming this then `sock_dtls_term_session` or `sock_dtls_terminate_session()` to avoid confusion.

Reply:
`sock_dtls_terminate_session()` sounds good.